backend/xivdpscalc/pps/sample_sch_rotation.py

```diff
@@ -0,0 +1,81 @@
+    This is not meant to be a perfect replication of a good sch rotation, this is primarily for testing and demonstration.
+    This operates by starting with a gcd 4 chain opener, and then using a rudimentary priority list to pick actions.
+    """
+
+    def __init__(self):
+        self.rotation = [SchAction.BIOLYSIS, SchAction.AETHERFLOW, SchAction.ENERGYDRAIN,
+                         SchAction.BROIL3, SchAction.BROIL3, SchAction.RUIN2, SchAction.CHAINSTRATAGEM, SchAction.ENERGYDRAIN,
+                         SchAction.BROIL3, SchAction.RUIN2, SchAction.ENERGYDRAIN, SchAction.SWIFTCAST, SchAction.BROIL3,
+                         SchAction.DISSIPATION, SchAction.ENERGYDRAIN, SchAction.BROIL3, SchAction.RUIN2, SchAction.ENERGYDRAIN]
+        self.index = 0
+
+    def get_action(self, current_time: ElapsedTime, cooldowns: dict[SchAction, float],
+                   remaining_effect_durations: dict[SchEffect, float], resources: dict[SchResource, int]) -> SchAction:
+        """
+        Returns a SchAction based on an arbitrary set of rules.
+        """
+        # broil by default
+        selected_action = SchAction.BROIL3
+        if self.index < len(self.rotation):
+            self.index += 1
+            selected_action = self.rotation[self.index - 1]
+        elif cooldowns[SchAction.BROIL3] >= 0.75:
+            # this is a weaving window
+            # the highest cooldown we can have without clipping assuming anim lock of 0.75 seconds
+            
+            # an dict of each ogcd with its usage constraint, ordered by priority
+            ogcd_constraints = {
+                SchAction.ENERGYDRAIN: resources[SchResource.AETHERFLOW] > 0, 
+                SchAction.CHAINSTRATAGEM: True,
+                SchAction.AETHERFLOW: resources[SchResource.AETHERFLOW] <= 0,
+                SchAction.DISSIPATION: resources[SchResource.AETHERFLOW] <= 0,
+                SchAction.SWIFTCAST: True}
+            
+            earliest_nonclip_cast_time = cooldowns[SchAction.BROIL3] - 0.75
+            for action in ogcd_constraints:
+                if cooldowns[action] <= earliest_nonclip_cast_time and ogcd_constraints[action]:
+                    selected_action = action
+                    break;
+        # if bio needs refreshing
+        elif remaining_effect_durations[SchEffect.BIOLYSIS] <= 0:
+            selected_action = SchAction.BIOLYSIS
+        # if chain stratagem, dissipation, or aetherflow are about to come up, but bio is still up, open a weaving window with r2
+        elif cooldowns[SchAction.CHAINSTRATAGEM] < 1.5 or \
+            ((cooldowns[SchAction.DISSIPATION] < 1.5) and resources[SchResource.AETHERFLOW] <= 1):
+            selected_action = SchAction.RUIN2
+        # if aetherflow or dissipation are coming up, we still have aetherflow reserves, and energy drain is up, r2 to ed
+        elif (cooldowns[SchAction.AETHERFLOW] < 30 or cooldowns[SchAction.DISSIPATION] < 30) and \
```

Comment:
what about swiftcast broil? where is that handled? in a non-dissipation minute you should be able to use 0 r2 casts because bio aetherflow ed (filler) bio ed swift broil ed (filler)

Reply:
i see that swiftcast is handled in the get_total_potency function which i find a bit odd, since it influences action selection. example being that you wouldn't want to select ruin 2 if you have swift buff active.